plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+			       fw_buf[fw_header_offset + NVM_VER_OFFSET_MAJOR],
+			       fw_buf[fw_header_offset + NVM_VER_OFFSET_MINOR]);
+	fw_vendor_id  = g_strdup_printf ("%02x%02x",
+			       fw_buf[fw_header_offset + NVM_VID_OFFSET_MAJOR],
+			       fw_buf[fw_header_offset + NVM_VID_OFFSET_MINOR]);
+	fw_product_id = g_strdup_printf ("%02x%02x",
+			       fw_buf[fw_header_offset + NVM_PID_OFFSET_MAJOR],
+			       fw_buf[fw_header_offset + NVM_PID_OFFSET_MINOR]);
+
+	g_debug ("writing Thunderbolt firmware version %s", fw_version);
+	g_debug ("writing Thunderbolt product-id %s"      , fw_product_id);
+	g_debug ("writing Thunderbolt vendor-id %s"       , fw_vendor_id);
+
+	/* compare vendor-id, product-id between firmware blob and NVM */
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, nvm_buf, NVM_READ_LENGTH,
+					     0, error)) {
+		g_prefix_error (error, "NVM READ error:");
+		return FALSE;
+	}
+	nvm_vendor_id  = g_strdup_printf ("%02x%02x", nvm_buf[NVM_VID_OFFSET_MAJOR],
+				                      nvm_buf[NVM_VID_OFFSET_MINOR]);
+	nvm_product_id = g_strdup_printf ("%02x%02x", nvm_buf[NVM_PID_OFFSET_MAJOR],
+				                      nvm_buf[NVM_PID_OFFSET_MINOR]);
+
+	if (((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) &&
+	    (g_strcmp0 (nvm_vendor_id , fw_vendor_id) != 0 ||
+	    g_strcmp0 (nvm_product_id, fw_product_id) != 0)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Thunderbolt firmware vendor_id %s, product_id %s"
+			     "doesn't match NVM vendor_id %s, product_id %s",
+			     fw_vendor_id, fw_product_id,
+			     nvm_vendor_id, nvm_product_id);
+		return FALSE;
+	}
+
+	/* firmware install */
+	fw_buf += fw_header_offset;
+	fw_blob_size -= fw_header_offset;
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, fw_buf, fw_blob_size, 0,
+					     error))
+		return FALSE;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
+	fu_device_set_version (device, fw_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_setup (FuDevice *device, GError **error)
+{
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *nvm_version = NULL;
+	g_autofree gchar *nvm_vendor_id  = NULL;
+	g_autofree gchar *nvm_product_id = NULL;
+	guchar buf[NVM_READ_LENGTH] = {0};
+
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, buf, NVM_READ_LENGTH, 0, error)) {
+		g_prefix_error (error, "NVM READ error:");
+		return FALSE;
+	}
+	nvm_version    = g_strdup_printf ("%02x.%02x", buf[NVM_VER_OFFSET_MAJOR],
+				                       buf[NVM_VER_OFFSET_MINOR]),
+	nvm_vendor_id  = g_strdup_printf ("%02x%02x" , buf[NVM_VID_OFFSET_MAJOR],
+				                       buf[NVM_VID_OFFSET_MINOR]);
+	nvm_product_id = g_strdup_printf ("%02x%02x" , buf[NVM_PID_OFFSET_MAJOR],
+				                       buf[NVM_PID_OFFSET_MINOR]);
+
+	/* only add known supported thunderbolt devices */
+	name = g_strdup_printf ("TBT-%s%s", nvm_vendor_id, nvm_product_id);
+	if (g_strcmp0 (name, DELL_DOCK_USB4_INSTANCE_ID) == 0)
```

Comment:
if this doesn't match, you should set an error and return false so it doesn't let anything else even setup.